src/server/models/order.ts

```diff
@@ -32,30 +32,105 @@ class Order extends Model<Order> {
-          console.log('cannot find order: creating one')
+        if (!orderCart) {
+          console.log(`cannot find cart for userId ${userId}: creating one`)
           cart = await Order.create({ userId, status: Status.Cart })
-        } else {
-          cart = orders.find(order => order.status === Status.Cart)
-          if (!cart) {
-            console.log('cannot find cart: creating one')
-            cart = await Order.create({ userId, status: Status.Cart })
-          }
-        }
+        } else cart = orderCart
+        // else {
+        //   cart = orders.find(order => order.status === Status.Cart)
+        //   if (!cart) {
+        //     console.log(`cannot find cart for userId : creating one`)
+        //     cart = await Order.create({ userId, status: Status.Cart })
+        //   }
+        // }
         return cart
       })
-      .then(cart => OrdersProducts.create({ orderId: cart.id, productId: productId }))
+      // 4. find existing ordersProduct instance and increase quantity by quantity.
+      // 5. if no existing ordersProduct instance, create instance.
+      .then(cart => {
+        // cart = order with status cart
+        return OrdersProducts.findAll({
+          where: {
+            orderId: cart.id
+          }
+        }).then(async (userOrderCart) => {
+          // userOrderCart is the orderProducts for the user's cart.
+          // If the productId exists, add quantity, else create product with that qty.
+          let foundProductFromCart = userOrderCart.filter(orderProdLineItem => orderProdLineItem.productId === productId)
+          if (foundProductFromCart.length === 0) {
+            // product does not exist in cart. add product to order
+            return await OrdersProducts.create({ orderId: cart.id, productId: productId, quantity: quantity })
+          } else {
+            // product exists in cart. increase quantity
+            let rowToUpdate = foundProductFromCart[0]
+            return await rowToUpdate.update({ quantity: rowToUpdate.quantity + quantity })
+          }
+        })
+      })
       .catch((e: Error) => console.log(`Failed to add to cart. \n${e}`))
   }
+
+  // Removes a product from cart. if optional paramter removeAll=true, then it removes all quantity of this item from the cart
+  static decreaseProductQuantityFromCart(userId: number, productId: number, removeAll: boolean = false) {
+    // find user's cart
+    return Order.findOne({
+      where: {
+        userId: userId,
+        status: Status.Cart
+      }
+    })
+      .then(cart => {
+        if (!cart) throw new Error(`User ${userId} has no cart. Should not empty a nonexistant cart.`)
+        OrdersProducts.findOne({
+          where: {
+            orderId: cart.id,
+            productId: productId
+          }
+        })
+          .then(async (orderProductLineItem) => {
+            if (!orderProductLineItem) throw new Error(`Product ${productId} does not exist in cart.`)
```

Comment:
I like this defensive coding... will definitely save us headaches when debugging